Add netcoreapp1.0 test target to Nancy.Tests

blairconrad opened the pull request:
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/NancyFx/Nancy/pulls) open
- [x] I have verified that I am following the Nancy [code style guidelines](https://github.com/NancyFx/Nancy/blob/45238076ad0b7f6ecabd6bae8469e30458d02efe/CONTRIBUTING.md#style-guidelines)
- [x] I have provided test coverage for my change (where applicable)

### Description

Supports #2612.

A preliminary attempt to convert Nancy.Tests to target a netcoreapp1.0.
I expect I've done many things wrong, and invite corrections.
I'll highlight what I think are the most egregious violations.

- update FakeItEasy to 3.0.0-alpha001
- add workaround so FakeItEasy gets required Castle.Core-beta002
- poorly patch the build so the project compiles and tests run

- currently
  - ~~`CookieBasedSessionsFixture.Should_load_valid_test_data` fails~~, and
  - ~~`DefaultNancyBootstrapperFixture.Container_should_ignore_specified_assemblies`
    is conditionally compiled out~~

dnfclas commented:
Hi __@blairconrad__, I'm your friendly neighborhood .NET Foundation Pull Request Bot (You can call me DNFBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla2.dotnetfoundation.org.
    </span>

TTYL, DNFBOT;

blairconrad commented on src/Nancy/Json/Simple/SimpleJson.cs:
```diff
@@ -1787,6 +1787,15 @@ public static IEnumerable<PropertyInfo> GetProperties(Type type)
+            return type.GetProperty(name);
+#endif
+        }
+
```
I'm not sure if this file is supposed to be modified or not…

blairconrad commented on test/Nancy.Tests/NoAppStartupsFixture.cs:
```diff
@@ -44,9 +44,13 @@ public IEnumerable<InstanceRegistration> InstanceRegistrations
 
         private static void ThrowWhenNoAppStartupsFixtureRuns()
         {
+#if CORE
```
I used `CORE` since I saw it in `Nancy/project.json`. Happy to change to something else. Or something elses.

blairconrad commented on test/Nancy.Tests/NoAppStartupsFixture.cs:
```diff
@@ -44,9 +44,13 @@ public IEnumerable<InstanceRegistration> InstanceRegistrations
         private static void ThrowWhenNoAppStartupsFixtureRuns()
         {
+#if CORE
+            if ( Environment.StackTrace.Contains("Nancy.Tests.NoAppStartupsFixture"))
```
less robust than `StackTrace().GetFrames()`.
Although if you're not worried about it, it might be cleaner to collapse to this.

blairconrad commented on test/Nancy.Tests/Resources/Link.Texts.Designer.cs:
```diff
@@ -10,8 +10,8 @@
     using System;
-    
-    
+    using Nancy.Extensions;
```
I am absolutely sure that editing the Designer files was a bad idea, but I didn't have any good ones. I'm all ears.

blairconrad commented on test/Nancy.Tests/Unit/DefaultNancyBootstrapperFixture.cs:
```diff
@@ -70,6 +70,7 @@ public void Request_should_be_available_to_request_startup()
             this.bootstrapper.RequestStartupLastRequest.ShouldBeSameAs(request);
         }
 
+#if !CORE
```
This was mostly laziness/expediency. I just haven't found the dynamic generator for the new framework yet.

blairconrad commented on test/Nancy.Tests/xUnitExtensions/UsingCultureAttribute.cs:
```diff
@@ -0,0 +1,49 @@
+#endif
+        }
+    }
+}
```
nicked from FakeItEasy

dnfclas commented:
__@blairconrad__, Thanks for signing the contribution license agreement so quickly! Actual humans will now validate the agreement and then evaluate the PR.
<br />Thanks, DNFBOT;

blairconrad commented on test/Nancy.Tests/project.json:
```diff
@@ -1,6 +1,8 @@
 {
+    "version": "1.4.1.0",
```
Without this, `CookieBasedSessionsFixture.Should_load_valid_test_data` fails because the `DefaultObjectSerializer` can't find Nancy.Tests, Version=1.4.1.0, Culture=neutral, PublicKeyToken=null, which was presumably encoded inside `CookieBasedSessionsFixture.ValidData`.
The full stack trace of the failure is

```
System.IO.FileLoadException : Could not load file or assembly 'Nancy.Tests, Version=1.4.1.0, Culture=neutral, PublicKeyToken=null'. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)
Stack Trace:
     at System.RuntimeTypeHandle.GetTypeByName(String name, Boolean throwOnError, Boolean ignoreCase, Boolean reflectionOnly, StackCrawlMarkHandle stackMark, IntPtr pPrivHostBinder, Boolean loadTypeFromPartialName, ObjectHandleOnStack type, ObjectHandleOnStack keepalive)
     at System.RuntimeTypeHandle.GetTypeByName(String name, Boolean throwOnError, Boolean ignoreCase, Boolean reflectionOnly, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean loadTypeFromPartialName)
     at System.RuntimeType.GetType(String typeName, Boolean throwOnError, Boolean ignoreCase, Boolean reflectionOnly, StackCrawlMark& stackMark)
     at System.Type.GetType(String typeName)
     at CallSite.Target(Closure , CallSite , Type , Object )
  src\Nancy\DefaultObjectSerializer.cs(69,0): at Nancy.DefaultObjectSerializer.Deserialize(String sourceString)
  src\Nancy\Session\CookieBasedSessions.cs(190,0): at Nancy.Session.CookieBasedSessions.Load(Request request)
  test\nancy.tests\Unit\Sessions\CookieBasedSessionsFixture.cs(370,0): at Nancy.Tests.Unit.Sessions.CookieBasedSessionsFixture.Should_return_blank_session_if_hmac_changed()
System.IO.FileLoadException : Could not load file or assembly 'Nancy.Tests, Version=1.4.1.0, Culture=neutral, PublicKeyToken=null'. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)
Stack Trace:
     at System.RuntimeTypeHandle.GetTypeByName(String name, Boolean throwOnError, Boolean ignoreCase, Boolean reflectionOnly, StackCrawlMarkHandle stackMark, IntPtr pPrivHostBinder, Boolean loadTypeFromPartialName, ObjectHandleOnStack type, ObjectHandleOnStack keepalive)
     at System.RuntimeTypeHandle.GetTypeByName(String name, Boolean throwOnError, Boolean ignoreCase, Boolean reflectionOnly, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean loadTypeFromPartialName)
     at System.RuntimeType.GetType(String typeName, Boolean throwOnError, Boolean ignoreCase, Boolean reflectionOnly, StackCrawlMark& stackMark)
     at System.Type.GetType(String typeName)
     at CallSite.Target(Closure , CallSite , Type , Object )
  src\Nancy\DefaultObjectSerializer.cs(69,0): at Nancy.DefaultObjectSerializer.Deserialize(String sourceString)
  src\Nancy\Session\CookieBasedSessions.cs(190,0): at Nancy.Session.CookieBasedSessions.Load(Request request)
  test\nancy.tests\Unit\Sessions\CookieBasedSessionsFixture.cs(355,0): at Nancy.Tests.Unit.Sessions.CookieBasedSessionsFixture.Should_load_valid_test_data()
```

This makes me a little concerned about the behaviour of `DefaultObjectSerializer`, especially if it's used to serialize and deserialize long-lived data, but I think that's outside the scope of this PR.

blairconrad replied:
Don't forget to look at this again. Instead of changing the version of this assembly, we could change the test data to the values @thomaslevesque suggested in https://github.com/NancyFx/Nancy/pull/2634#issuecomment-263126390.

Of course, either change would only fix the _test_. There's a larger issue of whether `DefaultObjectSerializer` should able to deserialize types into editions of those types from different-versioned assemblies, but that's beyond this PR. I could spin off an issue, if you like.

thecodejunkie replied:
Yeah this needs to be addressed separately. We cannot have one of our tests depend on a specific version. @jchannon

blairconrad replied:
I'm not entirely sure where to go from this comment. Are you suggesting to leave the code as is for now and then someone (you, me, @jchannon, a roving Samaritan) will "fix" the test separately?

thecodejunkie replied:
I'm suggesting that @jchannon takes a look since he implemented `DefaultObjectSerializer`. To be honest we changed the implementation, to it's current form, when we moved to core. The old implementation dependency on the binary formatter which was removed from .net core. Maybe it would even be wise to see if the binary formatter is back and make a call if we should revert back or not?

blairconrad replied:
Understood. The removal was a small pain point for us over at FakeItEasy too.
I thought I'd heard that binary serialization was coming back sometime, but I cannot find a reference now.

blairconrad commented:
With my latest change, `dotnet test` passes in Nancy.Tests on my machine. I see that Appveyor is less than happy for some reason.
I can try to investigate the other failures, but may need assistance from Team NancyFx.
Also, I've never set up a node to test `unix-x64` instances. I can give it a go later, but pointers would be appreciated. Or I can just add the framework to the project.json…

blairconrad commented:
Oh, I should've said. @thomaslevesque was a big help  in debugging the `Should_load_valid_test_data` failure today!

thomaslevesque commented:
FWIW, the correct values for `ValidData` and `ValidHmac` if the version remains "1.0.0.0" are:
```
private const string ValidData = "PzQmxxriR9Ht9a1rJMC1yAr5Ty+CIIE1/fXX/B2e32wtf8+KzGaBMOW6Ks9xb503haQnFTm9Z9QsJgoUPClwU6Ke25HRjXdKY7RQIG4XXT7APU/NV3KAJZuwbJObv22PR/yKbkTOEYAvn/m7FLM6jxvn5lCCw75Kw0vNvKiTbyE5ijh5EY4XfQHGt+5s6vrehh26reJBuXiY4hPwopGbOsvHNNw3HpbQPmut1qHiqX/w8naD2vuFmX0Dckv1Kkf+K3zG2BVHtXi+C3kufmUR+l/1C7p4Y7r+6n9+7o7Bf8bIDUCaqSgezFNO8qChVRJm4fanF+YkW7Oj/mKiYgMyEd8j0QQgnTIhf2QnXtdSILENyWTlvm94/QrqXfL1NqgmQ5+kiQJtDx30sxD+pOBUNaXoU7EL3AAmWpzSPCxDrKY=";
private const string ValidHmac = "uQvArgVpqmMxPyvoMqVI75u5yC4JXm+01PGQwDvrqaU=";
```



> This makes me a little concerned about the behaviour of DefaultObjectSerializer, especially if it's used to serialize and deserialize long-lived data

Me too. If I understand correctly, it means that deploying a new version of an app will always invalidate session cookies.

blairconrad commented:
Well, that's the AppVeyor build sorted.
The Travis I don't know about. When I add `unix-x64` to the `project.json`, I'm told:

```
λ dotnet restore
log  : Restoring packages for C:\PortableApps\Documents\Source\GitHub\Nancy\test\Nancy.Tests\project.json...
error: System.Diagnostics.Process 4.1.0 provides a compile-time reference assembly for System.Diagnostics.Process on .NETCoreApp,Version=v1.0, but there is no run-time assembly compatible with unix-x64.
error: System.IO.FileSystem.Watcher 4.0.0 provides a compile-time reference assembly for System.IO.FileSystem.Watcher on .NETCoreApp,Version=v1.0, but there is no run-time assembly compatible with unix-x64.
error: One or more packages are incompatible with .NETCoreApp,Version=v1.0 (unix-x64).
log  : Writing lock file to disk. Path: C:\PortableApps\Documents\Source\GitHub\Nancy\test\Nancy.Tests\project.lock.json
log  : C:\PortableApps\Documents\Source\GitHub\Nancy\test\Nancy.Tests\project.json
log  : Restore failed in 7155ms.

Errors in C:\PortableApps\Documents\Source\GitHub\Nancy\test\Nancy.Tests\project.json
    System.Diagnostics.Process 4.1.0 provides a compile-time reference assembly for System.Diagnostics.Process on .NETCoreApp,Version=v1.0, but there is no run-time assembly compatible with unix-x64.
    System.IO.FileSystem.Watcher 4.0.0 provides a compile-time reference assembly for System.IO.FileSystem.Watcher on .NETCoreApp,Version=v1.0, but there is no run-time assembly compatible with unix-x64.
    One or more packages are incompatible with .NETCoreApp,Version=v1.0 (unix-x64).
```

blairconrad commented:
Huzzah! Green builds. The netcoreapp1.0 Nancy.Tests aren't run on Travis (yet), but I'll look at that later. Then we can wrangle over details.

blairconrad commented:
Okay. We've a functional PR. Nancy.Tests has netcoreapp1.0 tests that run from the standard build script on both Windows and Unix (well, ubuntu at least).
Some of the changes are not pretty, but I'm more than happy to work to address anything the owners don't like.

blairconrad commented:
Hey, @khellang, @thecodejunkie, @jchannon, I'm bothering you specifically because I've seen you committing recently, so I know you're active.
I don't mean to be a grumpy potential contributor, but is there a pre-check that I've missed before a human gets involved in a PR?
It's been radio silence for a month and a day; long enough that I'm wondering whether it's me or it's you.
Thanks.

thecodejunkie commented:
@blairconrad sorry.. this is on my radar but I've been working on getting the `2.0.0-clinteastwood` release out and also preparing for Christmas (which is celebrated today in Sweden). On top of that my main focus has been to overhaul out build scripts, since I found them to be in a very inconsistent and messy state - all done now though. I'll shift my focus to this now since this is something that's very important to me. Can you give me a TL;DR of the changes that have been required, and what pain points you have come across?

blairconrad commented:
@thecodejunkie, no worries. I understand about competing priorities. I'm in no rush, and wouldn't've said anything except I couldn't even tell if the PR had been noticed.

Happy (belated) Christmas. I'll be celebrating on the 25th, so out of communication for a day or two.

I noticed the update to the build scripts. While resolving conflicts, I messed up your line endings, so just now repushed the last two commits.

TL;DR:

* added a netcoreapp1.0 framework section to `Nancy.Tests/project.json`. That's where all the trouble began :wink:
* most of conversion went pretty well, to get the Windows tests running. Standard "moving to .NET Core" stuff, as far as I can tell. I tried to make my updates fit the prevailing practices and style. I'm sure I failed somewhere. I'm more than happy to make any changes to fit in better, of course. Points of interest, may of which are indicated as comments in the PR:
  * changed dynamic assembly generation to use Roslyn
  * I had to set the Nancy.Tests version to 1.4.1 so the `DefaultObjectSerializer` could deal with the assembly
  * several Designer files used `Type.Assembly`, which isn't available in .NET Core. I changed to use `GetAssembly()`. Editing the files felt dirty, but I really don't know what else to do, short of excluding tests that rely on them
* Things were a little tougher on Unix. In the end I had to treat Nancy.Tests with the netcoreapp tests separately, so when we're on Unix, only running `DotNetCoreBuild` and `DotNetCoreRestore` on that project
  * To guard against the unlikely event of new test projects being added and not considered for netcoreapp treatment, I used a blacklist to filter out the projects we don't want to treat as DotNetCore. As we continue converting test projects (which I'm happy to work on, once we get this PR settled), the projects can be removed from the blacklist, which will eventually disappear.

thecodejunkie commented:
Crap.. I wrote my review comments a week ago and forgot to submit them =/ No wonder they all said `Pending` when I came back today to see if there'd been more commits =/

blairconrad commented:
> Crap.. I wrote my review comments a week ago and forgot to submit them

Ha! If I had a [nickel](https://en.wikipedia.org/wiki/Nickel_(Canadian_coin)) for every time I've done that, I'd have $0.15!

blairconrad commented on src/Nancy/Json/Simple/SimpleJson.cs:
```diff
@@ -1787,6 +1787,15 @@ public static IEnumerable<PropertyInfo> GetProperties(Type type)
 #endif
         }
 
+        public static PropertyInfo GetProperty(Type type, string name)
```
As it turns out, throwing a `.GetTypeInfo()` on the call site removed the need for this. Reverting.

blairconrad commented:
Uh, I'm a little confused by the AppVeyor build failure, @thecodejunkie. `test\Nancy.Tests\Unit\AppDomainAssemblyCatalogFixture.cs` isn't even a file in my branch (I really need to rebase sometime, but am afraid of losing your comments if I do so, so was hoping to address your comments first). Shall I rebase now instead?
Alternatively, I can continue on as I was, make you happy with the diffs, then rebase and see what trouble I get into!

thecodejunkie commented on test/Nancy.Tests/Unit/Bootstrapper/NancyBootstrapperBaseFixture.cs:
```diff
@@ -327,7 +327,7 @@ protected override INancyEngine GetEngineInternal()
             if (this.ShouldThrowWhenGettingEngine)
             {
-                throw new ApplicationException("Something when wrong when trying to compose the engine.");
+                throw new Exception("Something when wrong when trying to compose the engine.");
```
Na, fair enough. Let's use `Exception` since it's only in the context of the test itself

thecodejunkie commented on test/Nancy.Tests/Unit/Bootstrapper/NancyBootstrapperWithRequestContainerBaseFixture.cs:
```diff
@@ -245,7 +245,7 @@ protected override INancyEngine GetEngineInternal()
                 if (this.ShouldThrowWhenGettingEngine)
                 {
-                    throw new ApplicationException("Something when wrong when trying to compose the engine.");
+                    throw new Exception("Something when wrong when trying to compose the engine.");
```
Same as above

thecodejunkie commented on test/Nancy.Tests/Unit/Diagnostics/InteractiveDiagnosticsFixture.cs:
```diff
@@ -3,6 +3,7 @@
     using Nancy.Diagnostics;
     using System.Collections.Generic;
     using System.Linq;
+    using System.Reflection;
```
Why isn't that change shows in this change set? =)

blairconrad replied:
@thecodejunkie, you asked for me to point out any unresponded-to comments. I'm afraid you didn't see this one.

thecodejunkie replied:
Yep, that's fine 👍

thecodejunkie commented:
@blairconrad odd, that file has been there for a while https://github.com/NancyFx/Nancy/commits/master/test/Nancy.Tests/Unit/AppDomainAssemblyCatalogFixture.cs .. I think the best thing to do is to continue going though the review comments, ignore that AppVeyor is failing (for now) and once all the fixes have been applied you rebase against latest master (it's needed anyway). If it is still a problem after that, then we'll look at it again

blairconrad commented:
> odd, that file has been there for a while

On the one hand, I see what you're saying. It's been there for over 5 weeks. On the other hand, I had this PR ready for a full month before that, so by comparison, it's a baby! 😉 

I'm glad you feel the same way about continuing with the review and rebasing (and probably squashing later). I think it's the easier path.

I'll keep on with my comments and commits.

thecodejunkie commented:
> On the one hand, I see what you're saying. It's been there for over 5 weeks. On the other hand, I had this PR ready for a full month before that, so by comparison, it's a baby!

No the file was _moved_ 5 weeks ago. It's been there much longer, but when it was moved the history got lost

blairconrad commented on test/Nancy.Tests/Unit/DefaultNancyBootstrapperFixture.cs:
```diff
@@ -74,22 +80,32 @@ public void Request_should_be_available_to_request_startup()
+                    "Compilation failed:" + Environment.NewLine + "  " +
+                    string.Join(Environment.NewLine + "  ", emitResult.Diagnostics));
+
+#if CORE
```
Pulled out. I took some initiative deviated from your suggestion. Here was my thinking:

1. I didn't think an extension method was best, especially an `Assembly` exception. `MemoryStream` seemed the only option , and that didn't feel right.
2. There didn't seem to be a need to expose this functionality.

So I moved into an `AssemblyHelpers` instead. The location may not be correct - its sibling files contain only tests, not helpers, but I wasn't sure where to put it, and some other directories had mixtures. 

I realize I'm taking liberties, and am happy to change however you direct.

thecodejunkie commented on test/Nancy.Tests/Unit/ModelBinding/DefaultBodyDeserializers/JsonBodyDeserializerFixture.cs:
```diff
@@ -284,12 +285,11 @@ public void Should_Deserialize_Signed_And_Unsigned_Nullable_Numeric_Types()
 
 #if !MONO
         [Fact]
+        [UsingCulture("de-DE")]
```
Agreed

thecodejunkie commented on test/Nancy.Tests/Unit/NancyCookieFixture.cs:
```diff
@@ -37,25 +37,17 @@ public void Should_stringify_an_expiry_to_gmt_and_stupid_format()
         }
 
         [Fact]
+        [UsingCulture("fr-FR")]
```
Leave it

thecodejunkie commented on test/Nancy.Tests/Unit/Sessions/CookieBasedSessionsFixture.cs:
```diff
@@ -5,12 +5,12 @@ namespace Nancy.Tests.Unit.Sessions
 
     using Nancy.Bootstrapper;
     using Nancy.Cryptography;
+    using Nancy.Helpers;
```
Are you saying that tests, in this fixture, uses extension methods that came from `System.Web`, but since it's not on Core you added in `Nancy.Helpers` since it had matching extensions that could be used instead?

blairconrad replied:
Nearly. It's an entire missing class. `System.Web.HttpUtility`. My understanding is that `System.Web` is not available in .NET Core.
`Nancy.Helpers.HttpUtility` is used in the production code, e.g. in `NancyCookie`, so I took advantage here. Are you concerned that it weakens the tests?

thecodejunkie replied:
Na it's fine 👍

thecodejunkie commented:
Let me know if I missed to follow up on any of your comments

blairconrad commented:
@thecodejunkie, I rebased and added a commit to exclude `AppDomainAssemblyCatalogFixture`.
It turns out that `Directory.GetCurrentDirectory()` did not work under the full build. I've restored my change (and added a comment in the code to explain).
The odd thing is that [another test](https://ci.appveyor.com/project/NancyFx/nancy/build/2.0.0-ci0001840-qbqbokya#L1347) has failed on net452 in the AppVeyor build.
I've noticed tests of this ilk occasionally failing locally, even when I build from `master`. Is this known?

thecodejunkie commented:
@blairconrad think I'm caught up now.

> It turns out that `Directory.GetCurrentDirectory()` did not work under the full build. I've restored my change (and added a comment in the code to explain).

Where is that?

>The odd thing is that a test in DefaultRouteResolverFixture has failed on net452 in the AppVeyor build.
>I've noticed tests of its ilk occasionally failing locally, even when I build from master. Is this known?

Nope, first time I've seen that test fail

blairconrad commented:
> > It turns out that Directory.GetCurrentDirectory() did not work under the full build. I've restored my change (and added a comment in the code to explain).

> Where is that?

https://github.com/NancyFx/Nancy/pull/2634/commits/3f6624cff5579cb7d7d5f946f1032811ce0e95c5#diff-a433e18d53c53321ca83691eed103238R28

> > The odd thing is that a test in DefaultRouteResolverFixture has failed on net452 in the AppVeyor build.
> > I've noticed tests of its ilk occasionally failing locally, even when I build from master. Is this known?

> Nope, first time I've seen that test fail

Yeah, in `test/Nancy.Tests` on the `master` branch, run `dotnet -c Release` a few times. Generally I can see it after 2 or 3 builds.

blairconrad commented:
I think the failing `DefaultRouteResolverFixture` tests are the only outstanding point for this PR, no?
I've tried a little debugging, but not come up with any information yet. I can continue if you like, but I may not be the best equipped. Also, it may be worthwhile having someone from Team Nancy confirm (or deny) that it's an existing problem on master. If that's the case, perhaps the best course would be to spin off another issue?

thecodejunkie commented:
It's not a problem on `master` because if it were then we'd have seen it before on our CI servers that builds each PR and each new commit to `master`

blairconrad commented:
@thecodejunkie, at the risk of being fractious, I can't agree. Have you tried it as I suggested?

I see the failure on master on my personal machine. And I just now checked out master on my work computer (don't tell work) and got the same effect. Occasionally, not often, `Nancy.Tests.Unit.Routing.DefaultRouteResolverFixture` fails. On the work computer (a nicer machine than my home laptop, which may or may not be relevant) it took more runs, but after 6 or so, I got a failure:

```
D:\sandbox\Nancy\test\nancy.tests [master +0 ~19 -0]> dotnet test -c Release
Project Nancy (.NETFramework,Version=v4.5.2) was previously compiled. Skipping compilation.
Project Nancy.Authentication.Forms (.NETFramework,Version=v4.5.2) was previously compiled. Skipping compilation.
Project Nancy.Testing (.NETFramework,Version=v4.5.2) was previously compiled. Skipping compilation.
Project nancy.tests (.NETFramework,Version=v4.5.2) was previously compiled. Skipping compilation.
xUnit.net .NET CLI test runner (64-bit Desktop .NET win7-x64)
  Discovering: nancy.tests
  Discovered:  nancy.tests
  Starting:    nancy.tests
    Nancy.Tests.Unit.Routing.DefaultRouteResolverFixture.Should_resolve_single_capture(path: "/FOO/TESTING/PLOP", caseSensitive: False, expected: "TESTING") [FAIL]
      Assert.Equal() Failure
      Expected: Captured TESTING
      Actual:
      Stack Trace:
        Unit\Routing\DefaultRouteResolverFixture.cs(84,0): at Nancy.Tests.Unit.Routing.DefaultRouteResolverFixture.<Should_resolve_single_capture>d__4.MoveNext()
        --- End of stack trace from previous location where exception was thrown ---
           at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
           at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
        --- End of stack trace from previous location where exception was thrown ---
           at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
           at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
        --- End of stack trace from previous location where exception was thrown ---
           at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
           at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
  Finished:    nancy.tests
=== TEST EXECUTION SUMMARY ===
   nancy.tests  Total: 1825, Errors: 0, Failed: 1, Skipped: 0, Time: 6.416s
```

The very next run:

```
D:\sandbox\Nancy\test\nancy.tests [master +0 ~19 -0]> dotnet test -c Release
Project Nancy (.NETFramework,Version=v4.5.2) was previously compiled. Skipping compilation.
Project Nancy.Authentication.Forms (.NETFramework,Version=v4.5.2) was previously compiled. Skipping compilation.
Project Nancy.Testing (.NETFramework,Version=v4.5.2) was previously compiled. Skipping compilation.
Project nancy.tests (.NETFramework,Version=v4.5.2) was previously compiled. Skipping compilation.
xUnit.net .NET CLI test runner (64-bit Desktop .NET win7-x64)
  Discovering: nancy.tests
  Discovered:  nancy.tests
  Starting:    nancy.tests
    Nancy.Tests.Unit.RequestFixture.Should_respect_case_insensitivity_when_extracting_form_data_from_body_when_content_type_is_multipart_form_data [FAIL]
      Assert.Equal() Failure
      Expected: value,VALUE
      Actual:   value
      Stack Trace:
        Unit\RequestFixture.cs(330,0): at Nancy.Tests.Unit.RequestFixture.Should_respect_case_insensitivity_when_extracting_form_data_from_body_when_content_type_is_multipart_form_data()
    Nancy.Tests.Unit.Routing.DefaultRouteResolverFixture.Should_resolve_optional_capture_with_optional_specified(path: "/MOO/HOO/MOO", caseSensitive: False, expected: "HOO") [FAIL]
      Assert.Equal() Failure
      Expected: OptionalCapture HOO
      Actual:
      Stack Trace:
        Unit\Routing\DefaultRouteResolverFixture.cs(106,0): at Nancy.Tests.Unit.Routing.DefaultRouteResolverFixture.<Should_resolve_optional_capture_with_optional_specified>d__5.MoveNext()
        --- End of stack trace from previous location where exception was thrown ---
           at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
           at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
        --- End of stack trace from previous location where exception was thrown ---
           at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
           at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
        --- End of stack trace from previous location where exception was thrown ---
           at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
           at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
  Finished:    nancy.tests
=== TEST EXECUTION SUMMARY ===
   nancy.tests  Total: 1825, Errors: 0, Failed: 2, Skipped: 0, Time: 7.156s
```

Then 18 successes followed by 

```
Project Nancy (.NETFramework,Version=v4.5.2) was previously compiled. Skipping compilation.
Project Nancy.Authentication.Forms (.NETFramework,Version=v4.5.2) was previously compiled. Skipping compilation.
Project Nancy.Testing (.NETFramework,Version=v4.5.2) was previously compiled. Skipping compilation.
Project nancy.tests (.NETFramework,Version=v4.5.2) was previously compiled. Skipping compilation.
xUnit.net .NET CLI test runner (64-bit Desktop .NET win7-x64)
  Discovering: nancy.tests
  Discovered:  nancy.tests
  Starting:    nancy.tests
    Nancy.Tests.Unit.Routing.DefaultRouteResolverFixture.Should_resolve_optional_capture_with_optional_specified(path: "/MOO/HOO/MOO", caseSensitive: True, expected: "NA") [FAIL]
      Assert.Equal() Failure
      Expected: NotFound
      Actual:   OK
      Stack Trace:
        Unit\Routing\DefaultRouteResolverFixture.cs(106,0): at Nancy.Tests.Unit.Routing.DefaultRouteResolverFixture.<Should_resolve_optional_capture_with_optional_specified>d__5.MoveNext()
        --- End of stack trace from previous location where exception was thrown ---
           at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
           at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
        --- End of stack trace from previous location where exception was thrown ---
           at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
           at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
        --- End of stack trace from previous location where exception was thrown ---
           at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
           at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
  Finished:    nancy.tests
=== TEST EXECUTION SUMMARY ===
   nancy.tests  Total: 1825, Errors: 0, Failed: 1, Skipped: 0, Time: 6.940s
```

then one pass and

```
Project Nancy (.NETFramework,Version=v4.5.2) was previously compiled. Skipping compilation.
Project Nancy.Authentication.Forms (.NETFramework,Version=v4.5.2) was previously compiled. Skipping compilation.
Project Nancy.Testing (.NETFramework,Version=v4.5.2) was previously compiled. Skipping compilation.
Project nancy.tests (.NETFramework,Version=v4.5.2) was previously compiled. Skipping compilation.
xUnit.net .NET CLI test runner (64-bit Desktop .NET win7-x64)
  Discovering: nancy.tests
  Discovered:  nancy.tests
  Starting:    nancy.tests
    Nancy.Tests.Unit.Routing.DefaultRouteResolverFixture.Should_resolve_multi_literal(path: "/FOO/BAR/BAZ", caseSensitive: False) [FAIL]
      Assert.Equal() Failure
      Expected: MultipleLiteral
      Actual:
      Stack Trace:
        Unit\Routing\DefaultRouteResolverFixture.cs(62,0): at Nancy.Tests.Unit.Routing.DefaultRouteResolverFixture.<Should_resolve_multi_literal>d__3.MoveNext()
        --- End of stack trace from previous location where exception was thrown ---
           at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
           at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
        --- End of stack trace from previous location where exception was thrown ---
           at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
           at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
        --- End of stack trace from previous location where exception was thrown ---
           at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
           at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
  Finished:    nancy.tests
=== TEST EXECUTION SUMMARY ===
   nancy.tests  Total: 1825, Errors: 0, Failed: 1, Skipped: 0, Time: 5.974s
```

I think you've been lucky not to see the failure on the CI server.

blairconrad commented:
As you may have noticed, I've been running some experiments with the build in a place you can see.
After a few rebuilds of master, the [Nancy build did fail](https://ci.appveyor.com/project/NancyFx/nancy/build/2.0.0-ci0001840-unahedbd#L1397), but on a different test:
`Nancy.Testing.Tests.BrowserFixture.Should_return_JSON_serialized_querystring`.

Is this one known?

jchannon commented:
Not a known one no :(

On 8 February 2017 at 11:33, Blair Conrad <notifications@github.com> wrote:

> As you may have noticed, I've been running some experiments with the build
> in a place you can see.
> After a few rebuilds of master, the Nancy build did fail
> <https://ci.appveyor.com/project/NancyFx/nancy/build/2.0.0-ci0001840-unahedbd#L1397>,
> but on a different test:
> Nancy.Testing.Tests.BrowserFixture.Should_return_
> JSON_serialized_querystring.
>
> Is this one known?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NancyFx/Nancy/pull/2634#issuecomment-278303731>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAGapuivuDdlZaeeldDU8Yk66PXlIwrxks5raagKgaJpZM4K7CYa>
> .
>

blairconrad commented:
Thanks for the speedy confirmation, @jchannon. I'm not _trying_ to be difficult (it comes naturally), just trying to point out that maybe the tests have secrets.
Any objection to me continuing to try to reproduce on AppVeyor the failure that I see locally? It hogs the build a little, but I've never not seen it start right away, and I will settle down once I go to work. :smiley:

jchannon commented:
Nah go nuts! We REALLY appreciate your work here

On 8 February 2017 at 11:39, Blair Conrad <notifications@github.com> wrote:

> Thanks for the speedy confirmation, @jchannon
> <https://github.com/jchannon>. I'm not *trying* to be difficult (it comes
> naturally), just trying to point out that maybe the tests have secrets.
> Any objection to me continuing to try to reproduce on AppVeyor the failure
> that I see locally? It hogs the build a little, but I've never not seen it
> start right away, and I will settle down once I go to work. 😃
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NancyFx/Nancy/pull/2634#issuecomment-278304838>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAGapq0EvkpNL0XEmEK2l_qP1c55miVQks5raaltgaJpZM4K7CYa>
> .
>

xt0rted commented:
I think I've seen `Nancy.Testing.Tests.BrowserFixture.Should_return_JSON_serialized_querystring` fail locally. Even if it wasn't that one I have had tests fail locally that never fail on CI builds. My assumption was it was due to using the ReSharper test runner which is running them in parallel and how there are/were static config classes. If those tests don't reset the config defaults then any other test assuming they're getting the defaults could fail if run out of order. This was backed up by if I'd run the tests individually they'd pass, but running all of them they'd sometimes fail.

blairconrad commented:
Thanks for chiming in, @xt0rted. I even see the failures from the command-line build. Either using `build.ps1` or shortcutting and running a `dotnet test -c Release` on the relevant project.

jchannon commented:
try disabling parallel just in case

On 8 February 2017 at 14:57, Blair Conrad <notifications@github.com> wrote:

> Thanks for chiming in, @xt0rted <https://github.com/xt0rted>. I even see
> the failures from the command-line build. Either using build.ps1 or
> shortcutting and running a dotnet test -c Release on the relevant project.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NancyFx/Nancy/pull/2634#issuecomment-278351070>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAGapnKnRYPkFyejaFy4PppMsDGc_szZks5radftgaJpZM4K7CYa>
> .
>

xt0rted commented:
After looking through the commit history some more parallel tests were disabled in https://github.com/NancyFx/Nancy/pull/2259 and then enabled in https://github.com/NancyFx/Nancy/pull/2490

blairconrad commented:
It looks like the failing `DefaultRouteResolverFixture` tests were due to parallel testing and interference with the `StaticConfiguration` object.
I don't think there was anything else outstanding here, unless I've missed a comment.
@thecodejunkie, we need at least a rebase, and other than that the commits are a bit of a mess. I'm happy to squash them at the same time.
How's that sound?

blairconrad commented:
Oh, and we could potentially switch to FIE 3.0.0-rc001.

thecodejunkie commented:
sounds great =)

blairconrad commented:
Okay. There was a short delay while I made FIE 3.0.0-rc001, but I'll clean up the commits later today.

blairconrad commented:
3.0.0-rc001ed! Rebased! Squashed!

xt0rted commented:
https://github.com/NancyFx/Nancy/pull/2435#issuecomment-237790081 is where I had seen one of these tests failing.  It's old but `Should_return_JSON_serialized_form` was failing on TeamCity while passing locally. #2509 pertains to this as well as #2478 by the looks of it. I'm not sure if they're related anymore though.

jchannon commented:
@blairconrad This is EPIC! Thanks a million ❤️  👍

jchannon commented on build.cake:
```diff
@@ -255,6 +305,33 @@ Task("Test")
                 - GetFiles("./test/**/Nancy.ViewEngines.DotLiquid.Tests.xproj")
                 - GetFiles("./test/**/Nancy.Embedded.Tests.xproj"); //Embedded somehow doesnt get executed on Travis but nothing explicit sets it
+
+            var testProjects = GetFiles("./test/**/*.xproj");
```
Does this mean on unix `dotnet test` is called and then mono is called as well?

blairconrad replied:
Yeah. I figured that would be desirable. Can be changed.

blairconrad commented:
Hey, thanks for the merge, @jchannon. And to @thecodejunkie as well for the insightful comments during the review.

khellang commented:
:tada::heart:

blairconrad commented:
Embarrassingly, there's a warning generated when picking up the new FIE. I put the dependency as 3.0.0-rc001, not 3.0.0-rc001-build000097.

I'm considering converting another test project, if nobody objects. I can include the fix there, or as a separate PR.

blairconrad commented:
>  I can include the fix there

specifically #2705

thecodejunkie commented:
Awesome work @blairconrad ❤️  Thanks

blairconrad commented:
Glad to be of service, @thecodejunkie.
